File: backend/schemas/fund.py
"""
Pydantic schemas for API request/response validation.
"""
from datetime import date
from typing import Optional, List
from pydantic import BaseModel, Field


# ==================== Fund Filter Schemas ====================

class FundFilterRequest(BaseModel):
    """基金筛选请求 - 5-dimension cascade conditions."""
    # Basic Info
    fund_types: Optional[List[str]] = Field(default=None, description="基金类型列表 (25类)")
    setup_year_min: Optional[float] = Field(default=None, ge=0, description="成立年限最小值")
    setup_year_max: Optional[float] = Field(default=None, ge=0, description="成立年限最大值")
    scale_min: Optional[float] = Field(default=None, ge=0, description="规模最小值 (亿元)")
    scale_max: Optional[float] = Field(default=None, ge=0, description="规模最大值 (亿元)")
    company_names: Optional[List[str]] = Field(default=None, description="基金公司列表")
    
    # Holdings
    heavy_stocks: Optional[List[str]] = Field(default=None, description="重仓股票代码")
    heavy_bonds: Optional[List[str]] = Field(default=None, description="重仓债券代码")
    stock_position_min: Optional[float] = Field(default=None, ge=0, le=100, description="股票仓位最小值%")
    stock_position_max: Optional[float] = Field(default=None, ge=0, le=100, description="股票仓位最大值%")
    
    # Sector/Industry
    sw_level1: Optional[List[str]] = Field(default=None, description="申万一级行业")
    sw_level2: Optional[List[str]] = Field(default=None, description="申万二级行业")
    
    # Performance
    return_1y_min: Optional[float] = Field(default=None, description="近1年收益率最小值%")
    return_1y_max: Optional[float] = Field(default=None, description="近1年收益率最大值%")
    max_drawdown_1y_max: Optional[float] = Field(default=None, description="近1年最大回撤最大值%")
    sharpe_1y_min: Optional[float] = Field(default=None, description="近1年夏普最小值")
    
    # Manager Style
    manager_experience_min: Optional[float] = Field(default=None, description="经理从业年限最小值")
    turnover_rate_max: Optional[float] = Field(default=None, description="换手率最大值%")
    institution_holding_min: Optional[float] = Field(default=None, description="机构持仓占比最小值%")
    
    # Pagination
    page: int = Field(default=1, ge=1, description="页码")
    page_size: int = Field(default=50, ge=1, le=200, description="每页数量")
    
    # Sort
    sort_by: Optional[str] = Field(default="return_1y", description="排序字段")
    sort_order: str = Field(default="desc", pattern="^(asc|desc)$", description="排序方向")


class FundIndicatorResponse(BaseModel):
    """基金指标响应."""
    fund_code: str
    fund_name: str
    fund_type: str
    manager: Optional[str] = None
    setup_date: Optional[str] = None
    setup_year: Optional[float] = None
    scale: Optional[float] = None
    company_name: Optional[str] = None
    return_1y: Optional[float] = None
    volatility_1y: Optional[float] = None
    max_drawdown_1y: Optional[float] = None
    sharpe_1y: Optional[float] = None
    heavy_sector: Optional[str] = None


class FundFilterResponse(BaseModel):
    """基金筛选响应."""
    total: int
    page: int
    page_size: int
    funds: List[FundIndicatorResponse]


# ==================== Fund Compare Schemas ====================

class FundCompareRequest(BaseModel):
    """基金对比请求 - 最多15只."""
    fund_codes: List[str] = Field(..., min_length=1, max_length=15, description="基金代码列表")
    benchmark: str = Field(default="000300", description="基准指数代码 (默认沪深300)")
    start_date: Optional[str] = Field(default=None, description="开始日期 YYYY-MM-DD")
    end_date: Optional[str] = Field(default=None, description="结束日期 YYYY-MM-DD")


class CorrelationMatrixResponse(BaseModel):
    """相关性矩阵响应."""
    fund_codes: List[str]
    correlation_matrix: List[List[float]]  # N x N Pearson correlation
    calculation_date: str = Field(..., description="计算日期 YYYY-MM-DD")
    sample_size: int = Field(..., description="样本数量 (交易日数)")
    data_quality: Optional[dict] = Field(default=None, description="数据质量信息")


class FactorExposureResponse(BaseModel):
    """因子暴露响应 - 14 coefficients."""
    fund_code: str
    # 6 style factors
    large_cap_value: float
    large_cap_growth: float
    mid_cap_value: float
    mid_cap_growth: float
    small_cap_value: float
    small_cap_growth: float
    # 8 sector factors
    consumer: float
    tmt: float
    manufacturing: float
    healthcare: float
    cyclical: float
    utilities: float
    conglomerate: float
    cash: float
    intercept: float


# ==================== Market Overview Schemas ====================

class IndexQuoteResponse(BaseModel):
    """指数行情响应."""
    code: str
    name: str
    price: float
    change: float
    change_pct: float
    volume: Optional[float]
    amount: Optional[float]


class FearGreedResponse(BaseModel):
    """恐惧贪婪指数响应."""
    trade_date: str
    composite_score: float
    sentiment_status: str
    factor_volatility: Optional[float]
    factor_safe_haven: Optional[float]
    factor_margin_ratio: Optional[float]
    factor_volume_deviation: Optional[float]
    factor_futures_basis: Optional[float]
    factor_stock_strength: Optional[float]


class ERPSpreadResponse(BaseModel):
    """股债ERP响应."""
    index_code: str
    index_name: str
    trade_date: str
    pe_ttm: float
    treasury_yield_10y: float
    erp_spread: float
    percentile_rank_10y: Optional[float]
    index_close_price: Optional[float]


class TrajectoryPoint(BaseModel):
    """Phase space trajectory point."""
    x: float = Field(..., description="Crowding score (0-100)")
    y: float = Field(..., description="PE percentile (0-100)")
    date: str = Field(..., description="Date YYYY-MM-DD")


class TrajectoryVector(BaseModel):
    """Enhanced trajectory vector with velocity and rotation."""
    asset: str = Field(..., description="Asset code/name")
    start: TrajectoryPoint = Field(..., description="Start point T₀")
    end: TrajectoryPoint = Field(..., description="End point T₁")
    velocity: float = Field(..., description="Average velocity (crowding change per day)")
    rotation: str = Field(..., description="Rotation classification: clockwise/counter_clockwise/neutral/expansion/contraction")
    magnitude: Optional[float] = Field(None, description="Trajectory magnitude")
    direction_deg: Optional[float] = Field(None, description="Direction in degrees")


class TrajectoryResponse(BaseModel):
    """Phase space trajectory response for ECharts visualization."""
    vectors: List[TrajectoryVector] = Field(..., description="List of trajectory vectors")
    regime_change: bool = Field(..., description="Whether regime change detected")
    regime_change_count: Optional[int] = Field(None, description="Number of regime changes")


class CrowdingRotationVector(BaseModel):
    """拥挤度旋转向量 - Legacy format for backward compatibility."""
    asset_code: str
    asset_name: str
    t0_date: str
    t1_date: str
    t0_crowding: float
    t1_crowding: float
    t0_pe_percentile: float
    t1_pe_percentile: float


# ==================== Bond Market Schemas ====================

class YieldCurveResponse(BaseModel):
    """收益率曲线响应."""
    bond_type: str
    trade_date: str
    curve: List[dict]  # [{tenor: 1, yield: 2.5}, ...]


class CreditSpreadResponse(BaseModel):
    """信用利差响应."""
    bond_category: str
    rating: str
    tenor: str
    trade_date: str
    yield_ytm: float
    credit_spread: float
    percentile_rank: float


# ==================== Heatmap Matrix Schemas ====================

class HeatmapCell(BaseModel):
    """热力图单元格."""
    row: str  # 行标签 (如指数名)
    col: str  # 列标签 (如时间周期)
    value: float  # 数值
    color: str  # 颜色 rgba


class HeatmapMatrixResponse(BaseModel):
    """热力矩阵响应."""
    rows: List[str]
    cols: List[str]
    cells: List[HeatmapCell]


# ==================== Dashboard Aggregation Schemas ====================

class DashboardDataQuality(BaseModel):
    """Dashboard data quality indicator."""
    partial: bool = Field(..., description="Whether any sub-call failed")
    errors: Optional[dict] = Field(default=None, description="Error messages per metric")


class DashboardResponse(BaseModel):
    """首页看板聚合响应 - 恐惧贪婪、ERP、风格强度、拥挤度."""
    fear_greed: List[dict] = Field(default_factory=list, description="恐惧贪婪指数历史")
    erp: List[dict] = Field(default_factory=list, description="股债ERP历史")
    style_strength: List[dict] = Field(default_factory=list, description="风格强度历史")
    crowding: List[dict] = Field(default_factory=list, description="拥挤度分析历史")
    timestamp: str = Field(..., description="响应时间戳")
    data_quality: DashboardDataQuality = Field(..., description="数据质量信息")


# ==================== Fund Similarity Schemas ====================

class SimilarFund(BaseModel):
    """相似基金."""
    fund_code: str
    fund_name: str
    similarity: float = Field(..., description="Similarity score (0-1, higher is more similar)")


class FactorExposureItem(BaseModel):
    """因子暴露项."""
    factor_name: str
    factor_type: str = Field(..., description="style or sector")
    weight: float


class FundSimilarityResponse(BaseModel):
    """基金相似度响应."""
    fund_code: str
    similar_funds: List[SimilarFund] = Field(..., description="Top N similar funds")
    factor_exposure: List[FactorExposureItem] = Field(..., description="Factor exposure of input fund")
    calculation_method: str = Field(default="euclidean", description="Distance calculation method")
    elapsed_ms: Optional[float] = Field(None, description="Calculation time in milliseconds")


# ==================== AIP Calculator Schemas ====================

class AIPCalculateRequest(BaseModel):
    """定投计算请求."""
    fund_code: str = Field(..., description="基金代码")
    frequency: str = Field(..., pattern="^(weekly|biweekly|monthly)$", description="定投频率: weekly/biweekly/monthly")
    amount: float = Field(..., gt=0, description="每期投资金额(元)")
    start_date: str = Field(..., description="开始日期 YYYY-MM-DD")
    end_date: Optional[str] = Field(default=None, description="结束日期 YYYY-MM-DD, 默认今天")


class AIPCalculateResponse(BaseModel):
    """定投计算响应."""
    fund_code: str
    fund_name: str
    frequency: str
    amount: float
    total_investment: float = Field(..., description="累计投资金额")
    current_value: float = Field(..., description="当前市值")
    return_rate: float = Field(..., description="收益率(%)")
    max_drawdown: float = Field(..., description="最大回撤(%)")
    volatility: float = Field(..., description="年化波动率(%)")
    periods: int = Field(..., description="投资期数")
    units_total: float = Field(..., description="累计份额")
    lump_sum_comparison: dict = Field(..., description="一次性投资对比 {value, return_rate}")
    investment_dates: List[str] = Field(..., description="投资日期列表")
    nav_history: List[dict] = Field(..., description="净值历史 [{date, nav, units, value, cumulative_return}]")

File: frontend/src/api/fund.ts
/**
 * Fund API endpoints
 */
import api from './index'

// Types
export interface FundFilterParams {
  fund_types?: string[]
  setup_year_min?: number
  setup_year_max?: number
  scale_min?: number
  scale_max?: number
  company_names?: string[]
  return_1y_min?: number
  return_1y_max?: number
  max_drawdown_1y_max?: number
  sharpe_1y_min?: number
  new_high_ratio_min?: number
  page?: number
  page_size?: number
  sort_by?: string
  sort_order?: 'asc' | 'desc'
}

export interface FundItem {
  fund_code: string
  fund_name: string
  fund_type: string
  manager: string
  setup_date: string
  setup_year: number | null
  scale: number | null
  company_name: string
  return_1y: number | null
  volatility_1y: number | null
  max_drawdown_1y: number | null
  sharpe_1y: number | null
  heavy_sector: string | null
  new_high_ratio_1y?: number | null
  manager_honors?: string | null
}

export interface FundFilterResponse {
  total: number
  page: number
  page_size: number
  funds: FundItem[]
}

// Filter funds
export const filterFunds = async (params: FundFilterParams): Promise<FundFilterResponse> => {
  return api.post('/fund/filter', params)
}

// Get top gainers and losers
export const getTopFunds = async (limit = 10): Promise<{
  gainers: Array<{
    fund_code: string
    fund_name: string
    fund_type: string
    return_1y: number | null
  }>
  losers: Array<{
    fund_code: string
    fund_name: string
    fund_type: string
    return_1y: number | null
  }>
  timestamp: string
}> => {
  return api.get('/fund/top-funds', { params: { limit } })
}

// Get fund detail
export const getFundDetail = async (fundCode: string): Promise<FundItem> => {
  return api.get(`/fund/${fundCode}`)
}

// Compare funds
export const compareFunds = async (
  fundCodes: string[],
  benchmark = '000300'
): Promise<{ fund_codes: string[]; correlation_matrix: number[][] }> => {
  return api.post('/fund/compare', {
    fund_codes: fundCodes,
    benchmark,
  })
}

// Get fund issue calendar
export const getFundIssueCalendar = async (
  status?: string
): Promise<Array<{
  fund_code: string
  fund_name: string
  company: string
  subscribe_start_date: string
  subscribe_end_date: string
  status: string
  initial_scale: number | null
  delist_scale: number | null
}>> => {
  const params: Record<string, string> = {}
  if (status) params.status = status
  return api.get('/fund/issue', { params })
}

// Get fund companies
export const getFundCompanies = async (): Promise<Array<{
  company_id: string
  company_name: string
  establish_date: string | null
  total_scale: number | null
  non_money_scale: number | null
  fund_count: number | null
  manager_count: number | null
}>> => {
  return api.get('/fund/company')
}

// Get company distribution for treemap/bubble
export const getCompanyDistribution = async (
  companyId: string,
  distType: string = 'asset_class'
): Promise<{
  items: Array<{ item_name: string; weight: number }>
  is_simulated: boolean
}> => {
  return api.get(`/fund/company/${companyId}/distribution`, {
    params: { dist_type: distType }
  })
}

// Get fund NAV trend for sparkline
export const getFundNavTrend = async (
  fundCode: string,
  days: number = 30
): Promise<{
  nav_values: number[]
  dates: string[]
  is_simulated: boolean
}> => {
  return api.get(`/fund/${fundCode}/nav-trend`, {
    params: { days }
  })
}

// AIP (Automatic Investment Plan) Calculator Types
export interface AIPCalculateRequest {
  fund_code: string
  frequency: 'weekly' | 'biweekly' | 'monthly'
  amount: number
  start_date: string
  end_date?: string
}

export interface AIPCalculateResponse {
  fund_code: string
  fund_name: string
  frequency: string
  amount: number
  total_investment: number
  current_value: number
  return_rate: number
  max_drawdown: number
  volatility: number
  periods: number
  units_total: number
  lump_sum_comparison: { value: number; return_rate: number }
  investment_dates: string[]
  nav_history: Array<{
    date: string
    nav: number
    units: number
    value: number
    cumulative_return: number
  }>
}

// Calculate AIP
export const calculateAIP = async (
  data: AIPCalculateRequest
): Promise<AIPCalculateResponse> => {
  return api.post('/fund/aip-calculate', data)
}

// ==================== Holdings & Industry Types ====================

/**
 * Stock holding item in portfolio
 */
export interface HoldingItem {
  stock_code: string        // 股票代码
  stock_name: string        // 股票名称
  holding_ratio: number     // 持仓比例 (%)
  holding_value: number     // 持仓市值 (万元)
  change_direction: 'new' | 'increase' | 'decrease' | 'unchanged'  // 变动方向
}

/**
 * Industry allocation item
 */
export interface IndustryItem {
  industry_name: string     // 行业名称
  weight: number            // 占比 (%)
}

/**
 * Report date with holdings data
 */
export interface ReportDate {
  report_date: string       // 报告期 (YYYY-MM-DD)
  label: string             // 显示标签 (如 "2024年一季报")
}

/**
 * Holdings API response
 */
export interface HoldingsResponse {
  fund_code: string
  report_date: string
  report_dates: ReportDate[]
  holdings: HoldingItem[]
}

/**
 * Industry API response
 */
export interface IndustryResponse {
  fund_code: string
  report_date: string
  report_dates: ReportDate[]
  industries: IndustryItem[]
}

// Get fund holdings (top 10 stocks)
export const getFundHoldings = async (
  fundCode: string,
  reportDate?: string
): Promise<HoldingsResponse> => {
  const params: Record<string, string> = {}
  if (reportDate) params.report_date = reportDate
  return api.get(`/fund/${fundCode}/holdings`, { params })
}

// Get fund industry allocation
export const getFundIndustry = async (
  fundCode: string,
  reportDate?: string
): Promise<IndustryResponse> => {
  const params: Record<string, string> = {}
  if (reportDate) params.report_date = reportDate
  return api.get(`/fund/${fundCode}/industry`, { params })
}

// ==================== Stock Reverse Holding Types ====================

/**
 * Fund holding a specific stock
 */
export interface StockReverseHoldingItem {
  fund_code: string         // 基金代码
  fund_name: string         // 基金名称
  holding_ratio: number     // 持仓占净值比例
  holding_value?: number    // 持股市值 (万元)
  report_date: string       // 报告期
}

export interface StockReverseHoldingResponse {
  stock_code: string            // 股票代码
  stock_name: string            // 股票名称
  total_funds: number           // 持有该股票的基金总数
  aggregate_exposure: number    // 加总暴露度 (%)
  funds: StockReverseHoldingItem[]  // 持仓基金列表
}

export const getStockReverseHolding = async (
  stockCode: string,
  limit: number = 50
): Promise<StockReverseHoldingResponse> => {
  return api.get('/fund/stock-reverse', {
    params: { stock_code: stockCode, limit }
  })
}

export interface CrowdingAnalysisResponse {
  stock_code: string
  stock_name: string
  total_funds: number
  crowding_score: number
  hhi_index: number
  concentration_level: 'none' | 'low' | 'medium' | 'high' | 'extreme'
  overlap_coefficient: number
  avg_weight: number
  total_weight: number
  max_weight: number
  weight_std: number
  top_fund: string
  top_5_weight_pct: number
  quarter_distribution: Record<string, number>
}

export const getCrowdingAnalysis = async (
  stockCode: string
): Promise<CrowdingAnalysisResponse> => {
  return api.get(`/fund/stock-reverse/${stockCode}/crowding`)
}

export interface AggregationResponse {
  stock_code: string
  stock_name: string
  total_funds: number
  total_weight: number
  avg_weight: number
  max_weight: number
  weight_std: number
  top_fund: string
  quarter_distribution: Record<string, number>
}

export const getStockAggregation = async (
  stockCode: string
): Promise<AggregationResponse> => {
  return api.get(`/fund/stock-reverse/${stockCode}/aggregation`)
}

export const exportStockReverseHolding = async (
  stockCode: string,
  format: 'csv' | 'excel' = 'csv',
  limit: number = 100
): Promise<Blob> => {
  const response = await api.get(`/fund/stock-reverse/${stockCode}/export`, {
    params: { format, limit },
    responseType: 'blob'
  })
  return response
}
